### DISCUSSION TITLE
Custom Action :  unable to find version

### DISCUSSION THREAD

**Discussion by lgmorand** [community/community#39519] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi

I don't understand anymore how calling version works. I'm pretty sure it worked in the past.

I have a custom action with 3 [releases](https://github.com/lgmorand/github-action-hello/releases)

- v1.0.0 (based on tag v1.0.0)
- v1 (based on tag v1.0.1)
- v2 (based on tag v2.0.0)

If I call the action by the **exact** tag number, it [works perfectly](https://github.com/lgmorand/GHA-test-workflows/actions/runs/3504225441/workflow) but if I call it by the major version number (v1), it [does not work](https://github.com/lgmorand/GHA-test-workflows/actions/runs/3504220824/workflow), same goes for [v2](https://github.com/lgmorand/GHA-test-workflows/actions/runs/3504127161)

And the workflow can't resolve it: 
`Unable to resolve action `lgmorand/github-action-hello@v1`, unable to find version `v1``

I was expecting that by calling "v2", I would get the last version matching v2, including v2.0.0

What am I missing ?

**Comment by airtower-luna** [comment, marked as answer]
There's noting automatic, a version will only work if a ref (tag or branch) of the name exists. Some action repositories have branches (or moving tags) for major versions, some do not. If you want e.g. `v2` to work, you'll have to create such a ref and make sure it always points to the latest `v2.x` release.

**Reply by MichaelVoelkel** [reply]
But I see many repos where I can refer via @v1, it works and they don't have a v1 reference, at least no tag or branch?!
E.g. see https://github.com/dorny/test-reporter
What do I miss?

**Reply by airtower-luna** [reply]
You missed that that repository actually has a `v1` branch: https://github.com/dorny/test-reporter/tree/v1 Maybe you didn't look at the "stale" branches? :slightly_smiling_face:

**Reply by MichaelVoelkel** [reply]
Indeed! Stupid me, thank you

**Comment by lgmorand** [comment]
thanks @airtower-luna 

So if I understand you well, if I want people to always use the very "patch" version (i.e. v2.x), I need to update my "tag v2" to point to the last ref ? 

This way, they can call myaction@v2 and it would work ?

I'm asking because I thought that tags should be immutable but [here](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-tags-for-release-management), the doc says "move the tag version"

**Comment by lgmorand** [comment]
This article is pretty clear: https://dev.to/rob_bos/how-github-actions-versioning-system-works-lec

`So when you have a current release v4.2.5, you also need to re-tag that commit with the v4.2 tag, as well as the v4 tag. This way the runner can find the correct version to download.`

I think I was misguided because even GitHub does not follow best practices for its actions so I thought it would be implicit

**Reply by airtower-luna** [reply]
I agree it's misguided. :sweat_smile:

I'd probably make a v2 stable release *branch* instead of a tag if I wanted to make `...@v2` work, but:

* I'd recommend actually fixing the version (or even commit hash) and letting Depedabot take care of keeping it current. That avoids surprising changes, and potentially even malicious changes: If you review an action and pin the hash, you know it'll be exactly as reviewed, and you can review changes before updating. Where you trust the author using a specific minor version helps find any problems caused by new versions, because you'll *know* it was updated.
* I haven't checked if the GitHub Marketplace guidelines say anything about it (your action doesn't need to be listed there to be used, but it matters if you want it to be listed).

**Comment by lgmorand** [comment]
Agree for the hash version but for instance, when they announced the deprecation of NodeJS12, all actions started generating warning and I had to change around 100 workflows. it could have been transparent if everybody use semver properly :)

**Comment by airtower-luna** [comment]
Oh, that's definitely an annoying situation! That's why I mentioned Dependabot, you'd just have to accept the PRs (okay, 100 is still a lot! :grin:). And ideally tests would show you if an update might cause trouble before merging.
